wave_lang/kernel/wave/analysis/index_sequence_analysis.py

```diff
@@ -946,26 +949,109 @@ def apply_offset(node: fx.Node):
     trace.walk(apply_offset)
 
 
+def get_index(custom: CustomOp):
+    """Get the index from a CustomOp, handling special cases like MMABase."""
+    if not custom.indexing_dims:
+        return None
+    if isinstance(custom, MMABase):
+        return custom.acc.index
+    return custom.index
+
+
 def create_broadcast(
-    binary_op: BinaryPyOp,
-    to_broadcast: CustomOp,
+    op: CustomOp,
+    to_broadcast_arg_name: str,
     broadcast_dim: IndexSymbol,
     broadcast_size: int,
     target_node: CustomOp,
 ):
     """
-    Create a broadcast node for the given binary operator.
+    Create a broadcast node for any multi-operand operator.
     """
-    with binary_op.graph.inserting_before(binary_op.fx_node):
+    to_broadcast = get_custom(getattr(op, to_broadcast_arg_name))
+
+    with op.graph.inserting_before(op.fx_node):
         broadcasted = Broadcast(
             to_broadcast.fx_node, target_node.type.symbolic_shape
-        ).add_to_graph(binary_op.graph)
+        ).add_to_graph(op.graph)
         custom = get_custom(broadcasted)
-        custom.vector_shapes = binary_op.vector_shapes
+        custom.vector_shapes = op.vector_shapes
         custom.index = deepcopy(target_node.index)
         custom.index[broadcast_dim].size = broadcast_size
-        broadcast_idx = list(binary_op.node_args.values()).index(to_broadcast)
-        binary_op.update_arg(broadcast_idx, custom.fx_node)
+
+        op.update_arg(to_broadcast_arg_name, custom.fx_node)
+
+
+def resolve_broadcasting_for_op(
+    trace: CapturedTrace, op_type: type, operand_identifiers: list[str]
+):
+    """Generic broadcasting resolution for any multi-operand operation."""
+
+    ops = trace.walk(lambda node: isinstance(get_custom(node), op_type))
+
+    for op_node in ops:
+        custom = get_custom(op_node)
+
+        operands = []
+        for identifier in operand_identifiers:
+            operand_custom = get_custom(getattr(custom, identifier))
+
+            index = get_index(operand_custom)
+            dim, size = get_largest_index_and_size(index, operand_custom)
+            operands.append(
+                {"custom": operand_custom, "dim": dim, "size": size, "id": identifier}
+            )
+
+        sizes = [operand["size"] for operand in operands]
+        if len(set(sizes)) <= 1 or all(size == 1 for size in sizes):
+            continue
+
+        target = max(operands, key=lambda x: x["size"])
```

Comment:
Would also be nice to add back some of the context on when/why we deem some cases done :)